lib/vanagon/component/source.rb

```diff
@@ -1,78 +1,114 @@
+          end
         end
-      end
 
-      def self.string_rewrite(rule, url)
-        target_match = url.match(/.*\/([^\/]*)$/)
-        if target_match
-          target = target_match[1]
-          return File.join(rule, target)
-        else
-          raise Vanagon::Error, "Unable to apply url rewrite to '#{url}', expected to find at least one '/' in the url."
+        def parse_and_rewrite(uri)
+          url = URI.parse(uri)
+          return url unless url.scheme
+          rewrite(url.to_s, url.scheme)
         end
-      end
 
-      # Basic factory to hand back the correct {Vanagon::Component::Source} subtype to the component
-      #
-      # @param url [String] URL to the source (includes git@... style links)
-      # @param options [Hash] hash of the options needed for the subtype
-      # @param workdir [String] working directory to fetch the source into
-      # @return [Vanagon::Component::Source] the correct subtype for the given source
-      def self.source(url, options, workdir)
-        url_match = url.match(/^(.*)(@|:\/\/)(.*)$/)
-        uri_scheme = url_match[1] if url_match
-        local_source =  case uri_scheme
-                        when /^http/
-                          Vanagon::Component::Source::Http.new(self.rewrite(url, 'http'), options[:sum], workdir)
-                        when /^file/
-                          Vanagon::Component::Source::Local.new(self.rewrite(url, 'file'), workdir)
-                        when /^git/
-                          Vanagon::Component::Source::Git.new(self.rewrite(url, 'git'), options[:ref], workdir)
-                        else
-                          fail "Don't know how to handle source of type '#{uri_scheme}' from url: '#{url}'"
-                        end
+        # Needs a better name
+        def best_guess(uri, **options)
+          # First we git
+          if Vanagon::Component::Source::Git.valid_remote?(parse_and_rewrite(uri))
+            return Vanagon::Component::Source::Git.new parse_and_rewrite(uri),
+              sum: options[:sum],
+              workdir: options[:workdir]
+          end
+
+          # Then we HTTP
+          if Vanagon::Component::Source::Http.valid_url?(parse_and_rewrite(uri))
+            return Vanagon::Component::Source::Http.new parse_and_rewrite(uri),
+              sum: options[:sum],
+              workdir: options[:workdir]
+          end
+
+          # Then we try local
+          if Vanagon::Component::Source::Local.valid_file?(parse_and_rewrite(uri))
+            return Vanagon::Component::Source::Local.new parse_and_rewrite(uri),
+              workdir: options[:workdir]
+          end
 
-        return local_source
+          # Failing all of that, we give up
+          raise Vanagon::Error,
+            "Don't know how to handle unknown type: '#{uri}'"
+        end
+
+        # Needs a better name
+        def absolute_certainty(uri, type)
+          # try to parse the URI as whatever type was passed
```

Comment:
I'm going to add `:type` as a value for source definition, and this will be used there. I didn't realize I'd left the stub in when I committed this.